File: app/src/main/java/orinasa/njarasoa/maripanatokana/domain/model/Precipitation.kt
package orinasa.njarasoa.maripanatokana.domain.model

import java.util.Locale

/**
 * Value object encapsulating precipitation (rain/snow). Stores the canonical value in mm;
 * inches is always derived so both units are available once a value is set.
 */
@JvmInline
value class Precipitation private constructor(val mm: Double) {

    val inches: Double
        get() = mm * 0.03937

    fun displayMetric(): String = "%.1f mm".format(Locale.US, mm)

    fun displayImperial(): String = "%.2f in".format(Locale.US, inches)

    /** Dual-unit display: "2.5 mm / 0.10 in" */
    fun displayDual(): String = "${displayMetric()} / ${displayImperial()}"

    /** Returns (primary, secondary) based on preferred unit system. */
    fun displayDual(metricPrimary: Boolean): Pair<String, String> =
        if (metricPrimary) displayMetric() to displayImperial()
        else displayImperial() to displayMetric()

    companion object {
        fun fromMm(mm: Double): Precipitation = Precipitation(mm)

        fun fromInches(inches: Double): Precipitation = Precipitation(inches / 0.03937)
    }
}

File: react-native/src/domain/Precipitation.ts
/**
 * Value object encapsulating precipitation (rain/snow). Stores the canonical value in mm;
 * inches is always derived so both units are available once a value is set.
 */
export class Precipitation {
  readonly mm: number;

  private constructor(mm: number) {
    this.mm = mm;
  }

  get inches(): number {
    return this.mm * 0.03937;
  }

  displayMetric(): string {
    return `${this.mm.toFixed(1)} mm`;
  }

  displayImperial(): string {
    return `${this.inches.toFixed(2)} in`;
  }

  /** Returns [primary, secondary] based on preferred unit system. */
  displayDual(metricPrimary: boolean): [string, string] {
    return metricPrimary
      ? [this.displayMetric(), this.displayImperial()]
      : [this.displayImperial(), this.displayMetric()];
  }

  static fromMm(mm: number): Precipitation {
    return new Precipitation(mm);
  }

  static fromInches(inches: number): Precipitation {
    return new Precipitation(inches / 0.03937);
  }
}
